subsys/dfu/fmfu_fdev/src/fmfu_fdev.c

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <modem_update_decode.h>
+#include <drivers/flash.h>
+#include <logging/log.h>
+#include <dfu/fmfu_fdev.h>
+#include <nrf_fmfu.h>
+#include <mbedtls/sha256.h>
+
+LOG_MODULE_REGISTER(fmfu_fdev, CONFIG_FMFU_FDEV_LOG_LEVEL);
+
+#define MAX_META_LEN 1000
+
+static uint8_t meta_buf[1024];
+
+static int get_hash_from_flash(const struct device *fdev, size_t offset,
+			       size_t data_len, uint8_t *hash, uint8_t *buffer,
+			       size_t buffer_len)
+{
+	int err;
+	mbedtls_sha256_context sha256_ctx;
+	size_t end = offset + data_len;
+
+	mbedtls_sha256_init(&sha256_ctx);
+
+	err = mbedtls_sha256_starts_ret(&sha256_ctx, false);
+	if (err != 0) {
+		return err;
+	}
+
+	for (size_t p_offs = offset; p_offs < end; p_offs += buffer_len) {
+		size_t part_len = MIN(buffer_len, (end - p_offs));
+		int err = flash_read(fdev, p_offs, buffer, part_len);
+
+		if (err != 0) {
+			return err;
+		}
+
+		err = mbedtls_sha256_update_ret(&sha256_ctx, buffer, part_len);
+		if (err != 0) {
+			return err;
+		}
+	}
+
+	err = mbedtls_sha256_finish_ret(&sha256_ctx, hash);
+	if (err != 0) {
+		return err;
+	}
+
+	return 0;
+}
+
+static int load_segment(const struct device *fdev, size_t seg_size,
+			uint32_t seg_target_addr, uint32_t seg_offset,
+			uint8_t *buf)
+{
+	int err;
+	uint32_t read_addr = seg_offset;
+	size_t bytes_left = seg_size;
+
+	while (bytes_left) {
+		uint32_t read_len = MIN(NRF_FMFU_MODEM_BUFFER_SIZE, bytes_left);
+
+		err = flash_read(fdev, read_addr, buf, read_len);
+		if (err != 0) {
+			LOG_ERR("flash_read failed2: %d", err);
+			return err;
+		}
+
+		err = nrf_fmfu_memory_chunk_write(seg_target_addr, read_len,
+						  buf);
+		if (err != 0) {
+			LOG_ERR("nrf_fmfu_memory_chunk_write failed: %d", err);
+			return err;
+		}
+
+		LOG_DBG("wrote chunk: offset 0x%x target addr 0x%x size 0x%x",
+			read_addr, seg_target_addr, read_len);
+
+		seg_target_addr += read_len;
+		bytes_left -= read_len;
+		read_addr += read_len;
+	}
+
+	return 0;
+}
+
+static int prevalidate(uint8_t *buf, size_t wrapper_len)
+{
+	int err;
+
+	err = nrf_fmfu_verify_signature(wrapper_len, (void *)buf);
+	if (err != 0) {
+		LOG_ERR("nrf_fmfu_verify_signature failed: %d", errno);
+	}
+
+	return err;
+}
+
+static int load_segments(const struct device *fdev, uint8_t *meta_buf,
+			 size_t wrapper_len, const Segments_t *seg,
+			 size_t blob_offset, uint8_t *buf)
```

Comment:
IMO it is better to not provide init structs for static functions like this.

These functions are only defined to improve the readability of the function exposed through the public API.

The overhead of the struct param is not worth cleaning this up, especially since its internal only.

I do agree that this is not optimal though, and I tried to reduce the number of args, but the only way to do that was to introduce file scope variables, which I feel is worse than this.

Reply:
Yeah, that's totally fair. Works for me